### PR TITLE
Created registration endpoint | Added middleware function in middlewa…

### PR DESCRIPTION
**Pull request by IvayloSlavchev**
…re folder | Created tests for user input values

### CODE REVIEW

File: server/Authentication/authentication.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import { validateUserCredentials } from '../middlewares/registrationMiddleware'
+import client from '../database/database'
+
+const app = express();
+app.use(express.json());
+dotenv.config();
+
+//Test porpouses only
+app.get('/registration', (req: express.Request, res: express.Response) => {
+    const test = client.query('SELECT * FROM users');
```
Comment (petya0111) [inline server/Authentication/authentication.ts]: This query returns all users. But better filter some information fields of the user object to get fewer data. For ex. response of users must not return the password field.

File: server/Authentication/authentication.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    let hashingPasswordWithBcrypt = await bcrypt.hash(hashedPassword, 10);
+    hashedPassword = hashingPasswordWithBcrypt;
+
+
+
+    //Default values -> Should change in future
+    const account_enabled: boolean = true;
+    const account_locked: boolean = false;
+    const credentials_expired: boolean = false;
+    const created_at: string = '2023-01-20T22:00:00.000Z';
```
Comment (petya0111) [inline server/Authentication/authentication.ts]: Better use created at and updated at equalize automatic timestamp data. Or use `new Date()` to not use the hardcoded values on every registration.

**Review by petya0111 — CHANGES_REQUESTED** [review]
Some suggestions to think for dynamic data input fields.

File: server/Authentication/authentication.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    const test = client.query('SELECT * FROM users');
+    console.log(test);
+    res.status(200).send(test)
+})
+
+app.post('/registration', validateUserCredentials, async (req: express.Request, res: express.Response) => {
+    const { firstname, lastname, email, username, password, dayOfBirth, monthOfBirth, yearOfBirth } = req.body;
+
+    //Creating new variable in order to hash the password
+    let hashedPassword: string = password;
```
Comment (beval1) [inline server/Authentication/authentication.ts]: why create two variables for the same thing? one is enough

File: server/Authentication/authentication.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    //Default values -> Should change in future
+    const account_enabled: boolean = true;
+    const account_locked: boolean = false;
+    const credentials_expired: boolean = false;
+    const created_at: string = '2023-01-20T22:00:00.000Z';
+    const updated_at: string = '2023-01-20T22:00:00.000Z';
+
+    try {
+        await client.query(`INSERT INTO users(firstname, lastname, email, username, account_enabled, account_locked, credential_expired, dayOfBirth, monthOfBirth, yearOfBirth, created_at, updated_at, password) VALUES('${firstname}', '${lastname}', '${email}', '${username}', ${account_enabled}, ${account_locked}, ${credentials_expired}, ${dayOfBirth}, '${monthOfBirth}', ${yearOfBirth}, '${created_at}', '${updated_at}', '${hashedPassword}')`);
```
Comment (beval1) [inline server/Authentication/authentication.ts]: Wrong way of using INSERT statement. You should always parameterize your queries for security purposes.
see example here: https://dirask.com/posts/Node-js-PostgreSQL-Insert-query-DZXq2j

File: server/database/database.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+const client = new Client({
+    host: process.env.DATABASE_HOST,
+    user: process.env.DATABASE_USER,
+    password: process.env.DATABASE_PASSWORD,
+    database: process.env.DATABASE_NAME,
+}); 
+client.connect();
+
+client.query('SELECT * FROM users', (err, res) => {
```
Comment (beval1) [inline server/database/database.ts]: why is that here?

**Comment by beval1** [comment]
Here are few other problems:

Middleware is completely useless, you shouldn't check manually the type of every variable, leave this job to typescript.
Export the User model to another file and import it where needed
Keep logical seperation between the different layers of the application - those are web, database and service, check this article - https://martinfowler.com/bliki/PresentationDomainDataLayering.html

The code should be part of the auth folder, you have placed all the files in the server folder. 
You should place login and register endpoints in the same file in the auth folder (auth service)

File: server/Authentication/authentication.ts (hunk 4 above)
Comment (petya0111) [inline server/Authentication/authentication.ts]: 1. Here I can see that you don't check for an already presented username or email. Please check for an already registered user not to have duplicate data in DB.
2. Please make in the check for existing user - trim the fields and compare the two results with lowercasing them

**Review by petya0111 — CHANGES_REQUESTED** [review]
Check also for a registered user also @IvayloSlavchev

File: server/Authentication/authentication.ts (hunk 2 above)
Comment (IvayloSlavchev) [reply server/Authentication/authentication.ts]: Done

File: server/Authentication/authentication.ts (hunk 3 above)
Comment (IvayloSlavchev) [reply server/Authentication/authentication.ts]: Done

File: server/Authentication/authentication.ts (hunk 4 above)
Comment (IvayloSlavchev) [reply server/Authentication/authentication.ts]: Done

File: server/database/database.ts (hunk 5 above)
Comment (IvayloSlavchev) [reply server/database/database.ts]: Removed

File: server/Authentication/authentication.ts (hunk 1 above)
Comment (IvayloSlavchev) [reply server/Authentication/authentication.ts]: I actually remove it, because I was using it for test porpouses only.

**Review by petya0111 — APPROVED** [review]
Merged in the case of the history of the work in the server folder. ✔
